### PR TITLE
bluez: add disconnects hotfix

### LINKED ISSUE #117663: patch `bluez` to fix device disconnecting issues.
**Describe the bug**
`bluez` 5.56 has a [bug](https://github.com/bluez/bluez/issues/113) that causes some Bluetooth devices to disconnect frequently. Archlinux have already released a [hotfix](https://github.com/archlinux/svntogit-packages/commit/674c53af0aaa856e91c8fa73735cd520a312317c#diff-3e341d2d9c67be01819b25b25d5e53ea3cdf3a38d28846cda85a195eb9b7203a) to `bluez` 5.56.

**To Reproduce**
1. Upgrade to `bluez` 5.56
2. Connect to `JBL Go 2`, after a while it will disconnect at a random point.

**Additional context**

https://gitlab.freedesktop.org/pipewire/pipewire/-/issues/732#note_849446

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.106, NixOS, 21.05pre277486.f5e8bdd07d1 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - channels(root): `"nixos-21.05pre278406.d3f7e969b98"`
 - channels(iosmanthus): `"home-manager"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixos modules affected by the problem
module: hardware.bluetooth
```

### PR DESCRIPTION
**Pull request by ymatsiuk**
###### Motivation for this change
fixes #117663

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 117680` at 07d38576 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>141 packages marked as broken and skipped:</summary>
  <ul>
    <li>gnokii</li>
    <li>gnome3.gnome-books</li>
    <li>gnome3.gnome-documents</li>
    <li>gnunet-gtk</li>
    <li>gtkd</li>
    <li>libsForQt512.akonadi-calendar</li>
    <li>libsForQt512.akonadiconsole</li>
    <li>libsForQt512.alkimia</li>
    <li>libsForQt512.baloo-widgets</li>
    <li>libsForQt512.bomber</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>656 packages skipped due to time constraints:</summary>
  <ul>
    <li>adapta-gtk-theme</li>
    <li>adoptopenjdk-icedtea-web</li>
    <li>almanah</li>
    <li>apostrophe</li>
    <li>appgate-sdp</li>
    <li>aravis</li>
    <li>arc-theme</li>
    <li>arx-libertatis</li>
    <li>asciidoc-full</li>
    <li>asciidoc-full-with-plugins</li>
    <li>...</li>
  </ul>
</details>
<details>
  <summary>14 packages built successfully:</summary>
  <ul>
    <li>bluetooth_battery</li>
    <li>bluez (bluez5)</li>
    <li>bt-fw-converter</li>
    <li>cwiid</li>
    <li>obexfs</li>
    <li>obexftp</li>
    <li>openobex</li>
    <li>python38Packages.bleak</li>
    <li>python38Packages.bt_proximity</li>
    <li>python38Packages.pybluez</li>
    <li>python39Packages.bt_proximity</li>
    <li>python39Packages.ds4drv</li>
    <li>python39Packages.pybluez</li>
    <li>wiiuse</li>
  </ul>
</details>
<details>
  <summary>4 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/os-specific/linux/bluez/default.nix:119:3:
```
    |
119 |   meta = with lib; {
    |   ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/missing-patch-comment.md">missing-patch-comment</a>

Please add a comment on the line above, explaining the purpose of this patch.
Near pkgs/os-specific/linux/bluez/default.nix:48:5:
```
   |
48 |     ./28ddec8d6.patch #disconnect fix
   |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`gpl2` is a deprecated license, please check if project uses `gpl2Plus` or `gpl2Only` and change `meta.license` accordingly.

Near pkgs/os-specific/linux/bluez/default.nix:122:5:
```
    |
122 |     license = with licenses; [ gpl2 lgpl21 ];
    |     ^
```
</li>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/unclear-gpl.md">unclear-gpl</a>

`lgpl21` is a deprecated license, please check if project uses `lgpl21Plus` or `lgpl21Only` and change `meta.license` accordingly.

Near pkgs/os-specific/linux/bluez/default.nix:122:5:
```
    |
122 |     license = with licenses; [ gpl2 lgpl21 ];
    |     ^
```
</li>
  </ul>
</details>

File: pkgs/os-specific/linux/bluez/default.nix [hunk 1]
```diff
@@ -44,6 +44,11 @@ in stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ] ++ lib.optional doCheck "test";
 
+  patches = [
+    # disconnect fix -> https://github.com/bluez/bluez/commit/28ddec8d6
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/bluez/default.nix:48]:
```suggestion
    # disconnect fix -> https://github.com/bluez/bluez/commit/28ddec8d6b829e002fa268c07b71e4c564ba9e16
```

File: pkgs/os-specific/linux/bluez/default.nix [hunk 2]
```diff
@@ -44,6 +44,11 @@ in stdenv.mkDerivation rec {
...
+    ./28ddec8d6.patch
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/bluez/default.nix:49]: Can we fetch https://github.com/bluez/bluez/commit/28ddec8d6b829e002fa268c07b71e4c564ba9e16.patch instead?

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/bluez/default.nix (hunk 2 above)
Comment (ymatsiuk) [reply pkgs/os-specific/linux/bluez/default.nix:49]: I was actually looking into github docs on how to do it properly! Thanks.

File: pkgs/os-specific/linux/bluez/default.nix [hunk 3]
```diff
@@ -44,6 +45,12 @@ in stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ] ++ lib.optional doCheck "test";
 
+  patches = [(fetchpatch {
+    name = "disconnect-fix.patch";
+    url = "https://github.com/bluez/bluez/commit/28ddec8d6b829e002fa268c07b71e4c564ba9e16.patch";
+    sha256 = "sha256-vzMf1i44e4JrpL7cXbn9oDr+3B+Glf7dPW3QDstEnEM=";
+  })];
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/bluez/default.nix:52]:
```suggestion
  patches = [
    (fetchpatch {
      name = "disconnect-fix.patch";
      url = "https://github.com/bluez/bluez/commit/28ddec8d6b829e002fa268c07b71e4c564ba9e16.patch";
      sha256 = "sha256-vzMf1i44e4JrpL7cXbn9oDr+3B+Glf7dPW3QDstEnEM=";
    })
  ];
```

File: pkgs/os-specific/linux/bluez/default.nix [hunk 4]
```diff
@@ -44,6 +45,14 @@ in stdenv.mkDerivation rec {
 
   outputs = [ "out" "dev" ] ++ lib.optional doCheck "test";
 
+  patches = [
+    (fetchpatch {
```
Comment (worldofpeace) [inline pkgs/os-specific/linux/bluez/default.nix:49]:
```suggestion
    # Fixes https://github.com/NixOS/nixpkgs/issues/117663
    (fetchpatch {
```

**Review by worldofpeace — APPROVED** [review]
LGTM. whomever merges plz use squash merge

**Comment by Synthetica9** [comment]
> /rebase staging

whoa, that's a thing? 👀

**Comment by ymatsiuk** [comment]
shall I rebase?

**Comment by Synthetica9** [comment]
> shall I rebase?

No, the bot already did that, and I accidentally triggered it again 😓

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Pull request merged**